packages/dapp-svelte-wallet/api/src/date-now.js

```
@@ -0,0 +1,61 @@
+import { E } from '@agoric/eventual-send';
+import { observeNotifier } from '@agoric/notifier';
+import { makePromiseKit } from '@agoric/promise-kit';
+
+export const DEFAULT_TIMER_SERVICE_POLL_INTERVAL = 60_000n;
+export const DEFAULT_TIMER_DEVICE_SCALE = 1;
+
+export const makeTimerDeviceDateNow = async (
+  D,
+  timerDevice,
+  timerDeviceScale = DEFAULT_TIMER_DEVICE_SCALE,
+) => () => {
+  const stamp = D(timerDevice).getLastPolled();
+  const lastPolledStamp = parseInt(`${stamp}`, 10) * timerDeviceScale;
+  // console.error('synchronous time', stamp, nowStamp);
+  return lastPolledStamp;
+};
+
+export const makeTimerServiceDateNow = (
+  timerService,
+  timerPollInterval = DEFAULT_TIMER_SERVICE_POLL_INTERVAL,
+) => {
+  let lastPolledStamp = NaN;
+  const dateNow = () => lastPolledStamp;
+
+  const dateNowPK = makePromiseKit();
+
+  // Subscribe to a notifier that will poll the local timer service
+  // regularly.
+  const timerNotifier = E(timerService).makeNotifier(0n, timerPollInterval);
+
+  // Begin observing the notifier.
+  observeNotifier(timerNotifier, {
+    updateState(stamp) {
+      // console.error('asynchronous time', stamp);
+      lastPolledStamp = parseInt(`${stamp}`, 10);
+      dateNowPK.resolve(dateNow);
+    },
+  }).catch(e => dateNowPK.reject(e));
+
+  // Return the dateNow function after the notifier fires at least once.
+  return dateNowPK.promise;
+};
+
+export const makeDateNow = async ({
+  D,
+  timerDevice,
+  timerDeviceScale,
+  localTimerService,
+  localTimerPollInterval,
+}) => {
+  if (timerDevice) {
+    return makeTimerDeviceDateNow(D, timerDevice, timerDeviceScale);
+  }
+  if (localTimerService) {
+    return makeTimerServiceDateNow(localTimerService, localTimerPollInterval);
+  }
+
+  // No backend available.
```

What happens when this fails silently?

---

> What happens when this fails silently?

The `meta` property doesn't get timestamps.  Maybe the caller should print a warning in that case.